src/api/github/push.ts

```diff
@@ -0,0 +1,209 @@
+    messageBody,
+    fileChanges,
+    updateLocal,
+  });
+};
+
+export interface FileChanges {
+  additions: FileAddition[];
+  deletions: FileDeletion[];
+}
+
+/**
+ * Maps ChangedFiles to {@link https://docs.github.com/en/graphql/reference/input-objects#filechanges| FileChanges}
+ */
+export const mapChangedFilesToFileChanges = async (
+  changedFiles: ChangedFile[],
+): Promise<FileChanges> => {
+  const { added, deleted } = changedFiles.reduce<{
+    added: string[];
+    deleted: string[];
+  }>(
+    (files, changedFile) => {
+      const filePath = changedFile.path;
+      if (changedFile.state === 'deleted') {
+        files.deleted.push(filePath);
+      } else {
+        files.added.push(filePath);
+      }
+
+      return files;
+    },
+    { added: [], deleted: [] },
+  );
+
+  const additions: FileAddition[] = await Promise.all(
+    added.map(async (filePath) => ({
+      path: filePath,
+      contents: await fs.promises.readFile(filePath, {
+        encoding: 'base64',
+      }),
+    })),
+  );
+
+  const deletions: FileDeletion[] = deleted.map((filePath) => ({
+    path: filePath,
+  }));
+
+  return {
+    additions,
+    deletions,
+  };
+};
+
+interface CommitAndPushParams {
+  dir: string;
+  /**
+   * The branch name
+   */
+  branch: string;
+  /**
+   * The headline of the commit message
+   */
+  messageHeadline: string;
+  /**
+   * The body of the commit message
+   */
+  messageBody?: string;
+  /**
+   * File additions and deletions
+   */
+  fileChanges: FileChanges;
+  /**
+   * Updates the local git working directory to reflect the new remote state
+   */
+  updateLocal?: boolean;
+}
+
+/**
+ * Commits and pushes file changes up to a GitHub branch
+ */
+export const commitAndPush = async ({
+  dir,
+  branch,
+  messageHeadline,
+  messageBody,
+  fileChanges,
+  updateLocal = false,
+}: CommitAndPushParams): Promise<string> => {
+  const authToken = apiTokenFromEnvironment();
+  if (!authToken) {
+    throw new Error('Could not determine API token from the environment');
+  }
+
+  const [{ owner, repo }, headCommitId] = await Promise.all([
+    Git.getOwnerAndRepo({ dir }),
+    Git.getHeadCommitId({ dir }),
+  ]);
+
+  const input: CreateCommitOnBranchInput = {
+    branch: {
+      repositoryNameWithOwner: `${owner}/${repo}`,
+      branchName: branch,
+    },
+    message: {
+      headline: messageHeadline,
+      body: messageBody,
+    },
+    expectedHeadOid: headCommitId,
+    clientMutationId: 'skuba',
+    fileChanges,
+  };
+
+  const result = await graphql<CreateCommitResult>(
+    `
+      mutation Mutation($input: CreateCommitOnBranchInput!) {
+        createCommitOnBranch(input: $input) {
+          commit {
+            oid
+          }
+        }
+      }
+    `,
+    {
+      input,
+      headers: {
+        authorization: `bearer ${authToken}`,
+      },
+    },
+  );
+
+  if (updateLocal) {
+    await Promise.all(
+      [...fileChanges.additions, ...fileChanges.deletions].map((file) =>
```

Comment:
I'm a bit wary of how destructive this could be; it could blow away unstaged changes, which is arguably unintuitive given `commitAndPush` doesn't otherwise manipulate the local repo. Would this make more sense if this were exclusive to `commitAndPushAllChanges`?

Reply:
Hmmm I'm actually thinking maybe this should be called `push` given it doesn't really commit the changes.

Reply:
Hmm I guess it commits it to the remote..? 🤔

Reply:
Good point, `push` or `pushCommit` may make sense here 🤔